File: lib_arch/src/main/java/com/yyxnb/arch/annotations/BindRes.java
package com.yyxnb.arch.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/*
@Target指定Annotation用于修饰哪些程序元素。
@Target也包含一个名为”value“的成员变量，该value成员变量类型为ElementType[ ]，ElementType为枚举类型，值有如下几个：

ElementType.TYPE：能修饰类、接口或枚举类型
ElementType.FIELD：能修饰成员变量
ElementType.METHOD：能修饰方法
ElementType.PARAMETER：能修饰参数
ElementType.CONSTRUCTOR：能修饰构造器
ElementType.LOCAL_VARIABLE：能修饰局部变量
ElementType.ANNOTATION_TYPE：能修饰注解
ElementType.PACKAGE：能修饰包


RetentionPolicy.SOURCE	注解只保留在源文件，当Java文件编译成class文件的时候，注解被遗弃
RetentionPolicy.CLASS	注解被保留到class文件，但jvm加载class文件时候被遗弃，这是默认的生命周期
RetentionPolicy.RUNTIME	注解不仅被保存到class文件中，jvm加载class文件之后，仍然存在

 */

/**
 * 资源绑定
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface BindRes {

    /**
     * 布局id
     */
    int layoutRes() default 0;

    /**
     * 作用于Fragment判定为子页面，不处理状态栏
     */
    boolean subPage() default false;

    /**
     * 给系统窗口留出空间，xx屏
     */
    boolean fitsSystemWindows() default false;

    /**
     * 状态栏透明
     */
    boolean statusBarTranslucent() default true;

    /**
     * 状态栏文字颜色
     */
    @BarStyle int statusBarStyle() default BarStyle.DarkContent;

    /**
     * 状态栏颜色
     */
    int statusBarColor() default 0;

    /**
     * 侧滑 Fragment
     */
    @SwipeStyle int swipeBack() default SwipeStyle.Edge;

    /**
     * 页面是否需要登录
     */
    boolean needLogin() default false;

    /**
     * Activity是否作为容器
     */
    boolean isContainer() default false;

}

File: lib_arch/src/main/java/com/yyxnb/arch/annotations/BindRes.kt
package com.yyxnb.arch.annotations

/**
 * 资源绑定
 */
/*
@Target 作用域
定义该注解可使用的领域
AnnotationTarget.CLASS 可以给一个类型进行注解，类、接口、对象、甚至注解类本身
AnnotationTarget.ANNOTATION_CLASS 可以给一个注解类进行注解
AnnotationTarget.TYPE_PARAMETER 泛型参数（暂未支持）
AnnotationTarget.VALUE_PARAMETER 方法、构造函数的参数
AnnotationTarget.PROPERTY （计算）属性（该注解Java不可见）
AnnotationTarget.PROPERTY_GETTER 属性getter方法
AnnotationTarget.PROPERTY_SETTER 属性setter方法
AnnotationTarget.FIELD 字段变量，包括PROPERTY的备用字段（backing field）
AnnotationTarget.LOCAL_VARIABLE 局部变量
AnnotationTarget.CONSTRUCTOR 构造函数
AnnotationTarget.FUNCTION 方法、函数（不包括构造函数）
AnnotationTarget.FILE 文件整体
AnnotationTarget.TYPE 泛型支持
AnnotationTarget.TYPEALIAS typealias类型
AnnotationTarget.EXPRESSION 表达式
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
/*
@Retention 保留期
AnnotationRetention.SOURCE 注解只在源码阶段保留，在编译器进行编译时它将被丢弃忽视。 常用于代码检测。
AnnotationRetention.BINARY 注解被编译到二进制文件，但不会被加载到 JVM 中。 反射不可见。
AnnotationRetention.RUNTIME 注解被编译到二进制文件，它会被加载进入到 JVM 中，所以在程序运行时可以获取到它们。默认值。可用于运行时反射获取注解信息。
 */
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
//@MustBeDocumented 将注解中的元素包含到 Javadoc 中去。
@MustBeDocumented
//@Repeatable 该注解允许复数应用到同一目标（通过提供不同注解参数区分）
annotation class BindRes(
        /**
         * 布局id
         */
        val layoutRes: Int = 0,
        /**
         * 作用于Fragment判定为子页面，不处理状态栏
         */
        val subPage: Boolean = false,
        /**
         * 给系统窗口留出空间，xx屏
         */
        val fitsSystemWindows: Boolean = false,
        /**
         * 状态栏透明
         */
        val statusBarTranslucent: Boolean = true,
        /**
         * 状态栏文字颜色
         */
        val statusBarStyle: Int = BarStyle.DARK_CONTENT,
        /**
         * 状态栏颜色
         */
        val statusBarColor: Int = 0,
        /**
         * 侧滑 Fragment
         */
        val swipeBack: Int = SwipeStyle.EDGE,
        /**
         * 页面是否需要登录
         */
        val needLogin: Boolean = false,
        /**
         * Activity是否作为容器
         */
        val isContainer: Boolean = false
)